Apsitvarkom.DataAccess/PollutedLocationDTODatabaseRepository.cs

```diff
@@ -1,29 +1,47 @@
 ﻿using Apsitvarkom.Models;
 using Apsitvarkom.Models.DTO;
+using Apsitvarkom.Models.Mapping;
 using AutoMapper;
 using Microsoft.EntityFrameworkCore;
 
 namespace Apsitvarkom.DataAccess;
 
+/// <summary>
+/// Class for <see cref="PollutedLocationDTO" /> data handling using database.
+/// </summary>
 public class PollutedLocationDTODatabaseRepository : ILocationDTORepository<PollutedLocationDTO>
 {
-    private readonly PollutedLocationContext _context;
+    private readonly IPollutedLocationContext _context;
     private readonly IMapper _mapper;
 
-    public PollutedLocationDTODatabaseRepository(PollutedLocationContext context, IMapper mapper)
+    /// <summary>Constructor for the reader.</summary>
+    /// <param name="context">Database context for <see cref="PollutedLocation"/>.</param>
+    /// <param name="mapper">Mapper implementing profile <see cref="PollutedLocationProfile"/>.</param>
+    public PollutedLocationDTODatabaseRepository(IPollutedLocationContext context, IMapper mapper)
     {
         _context = context;
         _mapper = mapper;
     }
 
+    /// <inheritdoc />
     public async Task<IEnumerable<PollutedLocationDTO>> GetAllAsync()
     {
         var models = await _context.PollutedLocations.ToListAsync();
-
         return _mapper.Map<List<PollutedLocation>, IEnumerable<PollutedLocationDTO>>(models);
     }
 
-    public Task<IEnumerable<PollutedLocationDTO>> GetAllAsync(Location inRelationTo) => throw new NotImplementedException();
+    /// <inheritdoc />
+    public async Task<IEnumerable<PollutedLocationDTO>> GetAllAsync(Location inRelationTo)
+    {
+        return from pollutedLocation in await _context.PollutedLocations.ToListAsync()
+            orderby inRelationTo.DistanceTo(pollutedLocation)
+            select _mapper.Map<PollutedLocation, PollutedLocationDTO>(pollutedLocation);
+    }
 
-    public Task<PollutedLocationDTO?> GetByPropertyAsync(Func<PollutedLocationDTO, bool> propertyCondition) => throw new NotImplementedException();
+    /// <inheritdoc />
+    public async Task<PollutedLocationDTO?> GetByPropertyAsync(Func<PollutedLocationDTO, bool> propertyCondition)
+    {
+        var allLocations = await GetAllAsync();
```

Comment:
You should not depend on `GetAllAsync` here. AFAIK, EF's `DbSet` works by executing SQL queries when you need the result list (e.g. calling `ToListAsync`) or in other words when you move away from `IQueryable<T>` to a (for example) `List<T>`. So as a rule of thumb, you should keep `IQueryable<T>` as far as possible, because that will mean a more concrete SQL query.

So right now in this method this is happening: you are firstly executing SQL query that `SELECT`s all (by calling method `GetAllAsync`) and then doing `FirstOrDefault` on an `IEnumerable` (not `IQueryable`):

![image](https://user-images.githubusercontent.com/29711974/201526210-618db363-437e-4754-b990-2aca1d301660.png)

This is not optimal. Of course it works okay when we have a small number of rows in the database, but imagine we have a million of rows and this method would firstly query all of those rows and only then search for the first one. The optimal way would be to construct the SQL query to look for that one element immediately (without select all of the rows beforehand). You do that by chaining extension methods on `IQueryable`

Reply:
Another problem in our design here is that the `propertyCondition` is of type `Func<PollutedLocationDTO, bool>` but our `DbSet` is of type `PollutedLocation`, so you will also need to map between those types in the `FirstOrDefaultAsync` method.

We should think about this. I'm even more tempted to switch all of these repositories from DTO to business types and the controllers would just map the types themselves.